### PR TITLE
refactor: event consumer

### PR DESCRIPTION
**Pull request by adityathebe**

### CODE REVIEW

File: playbook/run_consumer.go [hunk 1]
```diff
@@ -0,0 +1,50 @@
...
+func NewEventQueueConsumer(db *gorm.DB, pool *pgxpool.Pool) *events.EventConsumer {
+	return events.NewEventConsumer(db, pool, "playbook_run_updates", EventConsumer).WithNumConsumers(5)
+}
+
+func EventConsumer(ctx *api.Context, batchSize int) error {
+	tx := ctx.DB().Begin()
+	if tx.Error != nil {
+		return fmt.Errorf("error initiating db tx: %w", tx.Error)
+	}
+	defer tx.Rollback()
```
Comment (yashmehrotra) [inline playbook/run_consumer.go:22]: Is this safe to run after doing a `tx.Commit()` ? does it fail silently in that case ?

File: events/event_queue.go [hunk 2]
```diff
@@ -0,0 +1,174 @@
...
+
+	for i := range allConsumers {
+		go allConsumers[i].Listen()
+	}
+}
+
+// newEventQueueConsumerFunc returns a new event consumer for the `event_queue` table
+// based on the given watch events and process batch function.
+func newEventQueueConsumerFunc(watchEvents []string, processBatchFunc ProcessBatchFunc) EventConsumerFunc {
+	return func(ctx *api.Context, batchSize int) error {
```
Comment (yashmehrotra) [inline events/event_queue.go:112]: Whats the advantage of returning this as a function vs calling the function directly ?

File: cmd/server.go [hunk 3]
```diff
@@ -231,13 +231,15 @@ var Serve = &cobra.Command{
 
 		go jobs.Start()
 
-		events.StartConsumers(db.Gorm, events.Config{
+		events.StartConsumers(db.Gorm, db.Pool, events.Config{
 			UpstreamPush: api.UpstreamConf,
 		})
 
-		playbookRunConsumer := playbook.NewQueueConsumer(db.Gorm, db.Pool)
+		playbookRunConsumer := playbook.NewEventQueueConsumer(db.Gorm, db.Pool)
```
Comment (yashmehrotra) [inline cmd/server.go:238]: Should we name his RunsConsumer or something with `run` in the name  ?

File: events/event_queue.go (hunk 2 above)
Comment (yashmehrotra) [reply events/event_queue.go:112]: Is it for passing a different consumer func to playbook runs?

File: playbook/run_consumer.go (hunk 1 above)
Comment (adityathebe) [reply playbook/run_consumer.go:22]: yea. Once it's committed, the rollback is a no-op.

File: events/event_queue.go (hunk 2 above)
Comment (adityathebe) [reply events/event_queue.go:112]: Ya, the `EventConsumer` needed to be generic and `watchEvents` and `processBatchFunc` are specific to consumers of `event_queue` and doesn't fit well with playbook even consumer.

File: cmd/server.go (hunk 3 above)
Comment (adityathebe) [reply cmd/server.go:238]: I see. And call the `.Listen` inside it?

Comment (yashmehrotra) [reply cmd/server.go:238]: yeah, something like that

**Comment by adityathebe** [comment]
@moshloop Can we merge this?

File: playbook/pg_listeners.go [hunk 4]
```diff
@@ -0,0 +1,89 @@
...
+		case id := <-pgNotifyPlaybookApprovalsInserted:
+			if err := onPlaybookRunNewApproval(ctx, id); err != nil {
+				logger.Errorf("%v", err)
+			}
+		}
+	}
+}
+
+// onApprovalUpdated is called when the playbook spec approval is updated
+func onApprovalUpdated(ctx *api.Context, playbookID string) error {
```
Comment (moshloop) [inline playbook/pg_listeners.go:46]: Is this the only place the approvals are managed? We can't use pg_notify as the only source as we will miss event when incident commander is down

Comment (adityathebe) [reply playbook/pg_listeners.go:46]: Right! I'll add new sync consumers but sync consumers are implemented in a later PR. I'll make the change there.

File: notification/notification_test.go [hunk 5]
```diff
@@ -104,13 +104,15 @@ var _ = ginkgo.Describe("Test Notification on incident creation", ginkgo.Ordered
 	})
 
 	ginkgo.It("should consume the event and send the notification", func() {
-		notifHandler := events.NewNotificationConsumer(db.Gorm)
-		sendHandler := events.NewNotificationSendConsumer(db.Gorm)
+		notifHandler := events.NewNotificationConsumer(db.Gorm, db.Pool)
```
Comment (moshloop) [inline notification/notification_test.go:107]:
```suggestion
		notificationHandler := events.NewNotificationConsumer(db.Gorm, db.Pool)
```

**Pull request merged**